src/PublicAllocator.sol

```diff
@@ -1,25 +1,126 @@
 // SPDX-License-Identifier: UNLICENSED
-pragma solidity ^0.8.13;
+pragma solidity 0.8.24;
 
-import {IMetaMorpho, MarketAllocation} from "../lib/metamorpho/src/interfaces/IMetaMorpho.sol";
+import {
+    Id, IMorpho, IMetaMorpho, MarketAllocation, MarketParams
+} from "../lib/metamorpho/src/interfaces/IMetaMorpho.sol";
 
-contract PublicAllocator {
+import {MarketParamsLib} from "../lib/metamorpho/lib/morpho-blue/src/libraries/MarketParamsLib.sol";
+
+import {MorphoLib} from "../lib/metamorpho/lib/morpho-blue/src/libraries/periphery/MorphoLib.sol";
+
+import {MorphoBalancesLib} from "../lib/metamorpho/lib/morpho-blue/src/libraries/periphery/MorphoBalancesLib.sol";
+
+import {SharesMathLib} from "../lib/metamorpho/lib/morpho-blue/src/libraries/SharesMathLib.sol";
+
+import {Market} from "../lib/metamorpho/lib/morpho-blue/src/interfaces/IMorpho.sol";
+import {UtilsLib} from "./libraries/UtilsLib.sol";
+import {Ownable2Step, Ownable} from "../lib/metamorpho/lib/openzeppelin-contracts/contracts/access/Ownable2Step.sol";
+
+import {ErrorsLib} from "./libraries/ErrorsLib.sol";
+import {EventsLib} from "./libraries/EventsLib.sol";
+import {FlowCap, FlowConfig, SupplyConfig, IPublicAllocatorStaticTyping} from "./interfaces/IPublicAllocator.sol";
+
+contract PublicAllocator is Ownable2Step, IPublicAllocatorStaticTyping {
+    using MorphoLib for IMorpho;
+    using MorphoBalancesLib for IMorpho;
+    using MarketParamsLib for MarketParams;
+    using SharesMathLib for uint256;
+    using UtilsLib for uint256;
+    using UtilsLib for uint128;
 
     /// STORAGE ///
 
-    address public owner;
-    IMetaMorpho public vault;
+    uint256 public fee;
+    IMetaMorpho public immutable VAULT;
+    IMorpho public immutable MORPHO;
+    mapping(Id => FlowCap) public flowCap;
+    mapping(Id => uint256) public supplyCap;
 
     /// CONSTRUCTOR ///
 
-    constructor(address newOwner, IMetaMorpho newVault) {
-        owner = newOwner;
-        vault = newVault;
+    constructor(address owner, address vault) Ownable(owner) {
+        if (vault == address(0)) {
+            revert ErrorsLib.ZeroAddress();
+        }
+        VAULT = IMetaMorpho(vault);
```

Comment:
> I'm not seeing what bad scenario it helps prevent?

It depends mostly on how you are using the factory's vault for and which are the requirements for these public allocators. 

If by specs you want to allow public allocators to interact with "official" vaults generated via the `MorphoFactory`, you need to add that check; otherwise you can leave it open to any vault.

I hope I was clear on what I meant :D

Reply:
I followed your suggestion in the `PublicAllocatorFactory` (41f205662d74590e436f1e4b548380ce7e77b970). That gives two tracks:
- With guardrails: you do everything through the factories, and if `publicAllocatorFactory.isPublicAllocator(account)`, then you know you're using the right contracts all the way down.
- Without guardrails: you can deploy your own `PublicAllocator` and connect it to your own `MetaMorpho`.